### ISSUE TITLE
📎 Implement `noUnnecessaryConditions`

### ISSUE THREAD

**Issue by arendjr** [biomejs/biome#6611]
With the latest improvements in type inference, implementing this rule should have become feasible.

### Discussed in https://github.com/biomejs/biome/discussions/6601

<div type='discussions-op-text'>

<sup>Originally posted by **iSchoen** June 27, 2025</sup>
There's a useful [rule](https://typescript-eslint.io/rules/no-unnecessary-condition/) in eslint that disallows unnecessary conditions.  So if a type indicates that an if statement or optional chain is not needed, then we get a helpful error.  Would be really cool to see in Biome.

```typescript
function head<T>(items: T[]) {
  // items can never be nullable, so this is unnecessary
  if (items) {
    return items[0].toUpperCase();
  }
}

function foo(arg: 'bar' | 'baz') {
  // arg is never nullable or empty string, so this is unnecessary
  if (arg) {
  }
}

function bar<T>(arg: string) {
  // arg can never be nullish, so ?. is unnecessary
  return arg?.length;
}

// Checks array predicate return types, where possible
[
  [1, 2],
  [3, 4],
].filter(t => t); // number[] is always truthy
```</div>

**Comment by vladimir-ivanov** [comment]
@arendjr what do you mean by `latest improvements in type inference` please?

ok, think I found what I need  - biome_js_type_info, please shout if I am on the wrong track.

**Comment by vladimir-ivanov** [comment]
tiny question - typescript rule is singular, do we want to name it noUnnecessaryCondition perhaps? or we make it plural deliberately? thanks

**Comment by arendjr** [comment]
Yes, indeed: You can query type information using the `Typed` service. Type information is inferred using `biome_js_type_info`, though it's stored as part of our module graph, which is inside `biome_module_graph`.

Since recently the type inference also has the ability to determine truthiness, falsiness and nullishness of types, which I think will be very helpful for this functionality.

> tiny question - typescript rule is singular, do we want to name it noUnnecessaryCondition perhaps? or we make it plural deliberately? thanks

We intentionally use plural form in our rule names, so it should be `noUnnecessaryConditions`.

**Comment by vladimir-ivanov** [comment]
unassigning from myself for now. Can come back to it later if no one else picks it up.

**Comment by ematipico** [comment]
The rule was implemented; however, there are still cases that need to be covered. cc @vladimir-ivanov to add the cases to cover here

**Comment by eliseumds** [comment]
I've detected three cases where it shouldn't error:

### Implicit boolean type

```ts
let myFlag = false;

window.addEventListener('message', (event) => {
  if (event.data.type === 'enable-my-flag') {
    myFlag = true;
  }
});

document.querySelector('#my-button')?.addEventListener('click', (event) => {
  if (myFlag) { // << it fails here, but it shouldn't
    // ...
  }
});
```

### Optional function argument with non-primitive types

```ts
function doSomething(user?: { name: string }) {
  if (user) { // << it fails here, but it shouldn't
    return user.name.toUpperCase();
  }

  return '__fallback__';
}
```

### Inside switch with string union

```ts
declare const myValue: 'a' | 'b' | 'c';

switch (myValue) { // << it fails here, but it shouldn't
  default:
    break;
}
```

**Comment by jackTabsCode** [comment]
```ts
while (true) {}
```

This also errors.

**Comment by ematipico** [comment]
> ```ts
> while (true) {}
> ```
> 
> This also errors.

I don't understand. It _should_ error

**Comment by ematipico** [comment]
@eliseumds 

Can you explain why  2 should error?

 In 2, the variable is undefined due to `users?`

Also, in 1, the variable mutates, so I need help understanding

**Comment by eliseumds** [comment]
> [@eliseumds](https://github.com/eliseumds)
> 
> Can you explain why 2 should error?
> 
> In 2, the variable is undefined due to `users?`
> 
> Also, in 1, the variable mutates, so I need help understanding

Oh these are examples of where we should *NOT* see errors.

Either way, [here is the same code in the typescript-eslint playground](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.ts&code=PTAEGUBUEECVNAYQIYGcCmoCMoC8%2BDCAoAG3QBdQBbATwDETkBzPUAM2RIwG4iiB3AJYA7ACYB7fgDpko0QFEAbumHkAMoNTkV6AE4AKAORV0qVM3SGANKH3plqgJR4AfKADeRUKEFtb9lXIpUWRyZClyGgAHTAJQQxVkACMyAFpaVLZGJkNnT29vWgZmVnJdAFd0Xm8AXyIax14iCQBjcpNVKQBHSt0acHQyFvJxA0MAYgyk8vIR4VyAfhk5JUCNLR0xlpJBFoBra1tnXDd8nz99Iuy80BBQAB570D1dUdAACz10LwK7qX%2BfnUGrw7vIAHIAESQaEwOEI8NwfCIdygcAQKAwoAATHgEQQiGxysJhoJxMJQBJwOITOR3iImPpyhhdAsAFweUDCZAmdlaXT00ANDw-Xy2Jl6G53R7PXSvXQfL4-by6CjlXTk8W6KRckwRcQAVSiMV0GPQ%2BkagL4ytV6viAH07RwSCQksh9g7DLw6qDIdDMTi8fi%2BCiYPA-ZgAMy4wPNdDbZAq0AtMlaag0ABqnEq7MMyEMoAAPvEkvmi4YWp6%2BKghOQWu9bLRMyRKpKwNKXm9PiqfqJ0BxyiRyKylaAkirkHsvciwOCoabQFHA-ggA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false) with `@typescript-eslint/no-unnecessary-condition` enabled.

**Comment by siketyan** [comment]
@ematipico 

> I don't understand. It should error

FYI, `typescript-eslint/no-unnecessary-condition` has an option to allow constant condition in loops as an exception: https://typescript-eslint.io/rules/no-unnecessary-condition/#allowconstantloopconditions

**Comment by DayKev** [comment]
The rule doesn't detect "always falsy" statements yet.
```ts
const a = [];

if (a) {} // correctly points out this is always truthy
if (!a) {} // incorrectly does not point out this is always falsy
```
https://biomejs.dev/playground/?lineWidth=120&indentStyle=space&lintRules=noUnnecessaryConditions&analyzerFixMode=safeAndUnsafeFixes&code=YwBvAG4AcwB0ACAAYQAgAD0AIABbAF0AOwAKAAoAaQBmACAAKABhACkAIAB7AH0AIAAvAC8AIABjAG8AcgByAGUAYwB0AGwAeQAgAHAAbwBpAG4AdABzACAAbwB1AHQAIAB0AGgAaQBzACAAaQBzACAAYQBsAHcAYQB5AHMAIAB0AHIAdQB0AGgAeQAKAGkAZgAgACgAIQBhACkAIAB7AH0AIAAvAC8AIABpAG4AYwBvAHIAcgBlAGMAdABsAHkAIABkAG8AZQBzACAAbgBvAHQAIABwAG8AaQBuAHQAIABvAHUAdAAgAHQAaABpAHMAIABpAHMAIABhAGwAdwBhAHkAcwAgAGYAYQBsAHMAeQA%3D&language=ts

The TypeScript-ESLint rule does catch both correctly: https://typescript-eslint.io/play/#ts=5.9.2&fileType=.ts&code=MYewdgzgLgBAhjAvDA2gXQNwFgBQuCWAZjABRwCUMA3jAL4wD0DMARviALYCmMwcUwABZcIMKIPwQCxEgEIK1Oo2ZtOPACYgRMMCFh8BgsRIhA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

**Comment by alex-dev** [comment]
```ts
export function test(value: "a" | "b" | "c") {
  switch (value) {
    case "a":
      return 1;
    case "b":
      return 2;
    case "c":
      return 3;
  }
}
```
This reports `lint/nursery/noUnnecessaryConditions` on the `switch` statement.

**Comment by jpike88** [comment]
```
function bar(arg: string) {
	return arg?.length;
}
```

This case doesn't error (unnecessary optional), but it should.

**Comment by jpike88** [comment]
And one more case that doesn't error when it should (arg.test should always be present which makes the `||` unnecessary):

```
interface Foo {
  test: string[]
}

function bar(arg: Foo) {
    return arg.test || [];
}
```

note: if arg itself is a `string[]`, then `arg || []` errors which is correct. so there's something weird going on when we're looking at the property of an interface.

**Comment by leevare** [comment]
I don't think this should report an error.

```ts
const value = Math.random();

switch (true) {  // <- An error occurred here.
  case value > 0.5:
    // ...
    break;
  case value > 0.1:
    // ...
    break;
  default:
    // ...
    break;
}
```

https://biomejs.dev/playground/?lintRules=noUnnecessaryConditions&code=YwBvAG4AcwB0ACAAdgBhAGwAdQBlACAAPQAgAE0AYQB0AGgALgByAGEAbgBkAG8AbQAoACkAOwAKAAoAcwB3AGkAdABjAGgAIAAoAHQAcgB1AGUAKQAgAHsACgAgACAAYwBhAHMAZQAgAHYAYQBsAHUAZQAgAD4AIAAwAC4ANQA6AAoAIAAgACAAIAAvAC8AIAAuAC4ALgAKACAAIAAgACAAYgByAGUAYQBrADsACgAgACAAYwBhAHMAZQAgAHYAYQBsAHUAZQAgAD4AIAAwAC4AMQA6AAoAIAAgACAAIAAvAC8AIAAuAC4ALgAKACAAIAAgACAAYgByAGUAYQBrADsACgAgACAAZABlAGYAYQB1AGwAdAA6AAoAIAAgACAAIAAvAC8AIAAuAC4ALgAKACAAIAAgACAAYgByAGUAYQBrADsACgB9AA%3D%3D&language=ts

**Comment by DayKev** [comment]
> I don't think this should report an error.
> 
> const value = Math.random();
> 
> switch (true) {
>   case value > 0.5:
>     // ...
>     break;
>   case value > 0.1:
>     // ...
>     break;
>   default:
>     // ...
>     break;
> }

Well, you should be using `if/else if/else` statements instead, then you won't get an error. `switch (true)` should definitely emit a diagnostic.

**Comment by hilja** [comment]
> Well, you should be using if/else if/else statements instead, then you won't get an error. switch (true) should definitely emit a diagnostic.

Passing true to switch case is an old trick, some people prefer it. Also TypeScript has [switch true narrowing](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-3.html#switch-true-narrowing).

Some people understandably don't like it, this almost could be it's own rule, or `noUnnecessaryConditions` could have a setting for it.

**Comment by hilja** [comment]
Optional param ignored if not inline type:

```ts
type Args = 'string'

function foo(args?: Args) {
  // Error
  if (args) console.log(args)
}

function bar(args?: string) {
  // Works
  if (args) console.log(args)
}
```

[Playground](https://biomejs.dev/playground/?lintRules=noUnnecessaryConditions&code=dAB5AHAAZQAgAEEAcgBnAHMAIAA9ACAAJwBzAHQAcgBpAG4AZwAnAAoACgBmAHUAbgBjAHQAaQBvAG4AIABmAG8AbwAoAGEAcgBnAHMAPwA6ACAAQQByAGcAcwApACAAewAKACAAIABpAGYAIAAoAGEAcgBnAHMAKQAgAGMAbwBuAHMAbwBsAGUALgBsAG8AZwAoAGEAcgBnAHMAKQAKAH0ACgAKAC8ALwAgAFcAbwByAGsAcwAKAGYAdQBuAGMAdABpAG8AbgAgAGIAYQByACgAYQByAGcAcwA%2FADoAIABzAHQAcgBpAG4AZwApACAAewAKACAAIABpAGYAIAAoAGEAcgBnAHMAKQAgAGMAbwBuAHMAbwBsAGUALgBsAG8AZwAoAGEAcgBnAHMAKQAKAH0A&language=ts)

**Comment by jackTabsCode** [comment]
> Optional param ignored if not inline type:
> 
> type Args = 'string'
> 
> function foo(args?: Args) {
>   // Error
>   if (args) console.log(args)
> }
> 
> function bar(args?: string) {
>   // Works
>   if (args) console.log(args)
> }
> 
> [Playground](https://biomejs.dev/playground/?lintRules=noUnnecessaryConditions&code=dAB5AHAAZQAgAEEAcgBnAHMAIAA9ACAAJwBzAHQAcgBpAG4AZwAnAAoACgBmAHUAbgBjAHQAaQBvAG4AIABmAG8AbwAoAGEAcgBnAHMAPwA6ACAAQQByAGcAcwApACAAewAKACAAIABpAGYAIAAoAGEAcgBnAHMAKQAgAGMAbwBuAHMAbwBsAGUALgBsAG8AZwAoAGEAcgBnAHMAKQAKAH0ACgAKAC8ALwAgAFcAbwByAGsAcwAKAGYAdQBuAGMAdABpAG8AbgAgAGIAYQByACgAYQByAGcAcwA%2FADoAIABzAHQAcgBpAG4AZwApACAAewAKACAAIABpAGYAIAAoAGEAcgBnAHMAKQAgAGMAbwBuAHMAbwBsAGUALgBsAG8AZwAoAGEAcgBnAHMAKQAKAH0A&language=ts)

This errors for me:
```ts
function fn(nullable?: object) {
       // Error
	if (nullable) {
		
	}
}
```

**Comment by joprice** [comment]
The eslint rule also fails when comparing null to undefined. This is an important one for correctness. Should this rule be augmented to do that or is there another rule that can enforce this? Example:

```ts
const x: undefined | string = undefined;
if (x === null) {
}
```

triggers:

```
error  Unnecessary conditional, comparison is always false, since `undefined === null` is false  @typescript-eslint/no-unnecessary-condition
```

**Comment by arendjr** [comment]
@joprice I think that may be a useful improvement for this rule indeed, especially if it brings parity with ESLint.

**Comment by rtaque** [comment]
> > Optional param ignored if not inline type:
> > type Args = 'string'
> > function foo(args?: Args) {
> > // Error
> > if (args) console.log(args)
> > }
> > function bar(args?: string) {
> > // Works
> > if (args) console.log(args)
> > }
> > [Playground](https://biomejs.dev/playground/?lintRules=noUnnecessaryConditions&code=dAB5AHAAZQAgAEEAcgBnAHMAIAA9ACAAJwBzAHQAcgBpAG4AZwAnAAoACgBmAHUAbgBjAHQAaQBvAG4AIABmAG8AbwAoAGEAcgBnAHMAPwA6ACAAQQByAGcAcwApACAAewAKACAAIABpAGYAIAAoAGEAcgBnAHMAKQAgAGMAbwBuAHMAbwBsAGUALgBsAG8AZwAoAGEAcgBnAHMAKQAKAH0ACgAKAC8ALwAgAFcAbwByAGsAcwAKAGYAdQBuAGMAdABpAG8AbgAgAGIAYQByACgAYQByAGcAcwA%2FADoAIABzAHQAcgBpAG4AZwApACAAewAKACAAIABpAGYAIAAoAGEAcgBnAHMAKQAgAGMAbwBuAHMAbwBsAGUALgBsAG8AZwAoAGEAcgBnAHMAKQAKAH0A&language=ts)
> 
> This errors for me:
> 
> function fn(nullable?: object) {
>        // Error
> 	if (nullable) {
> 		
> 	}
> }

Seems to me that only primitive type parameters are correctly infered as optional

**Comment by arendjr** [comment]
@rtaque that shouldn't be the case, but it's been a while since I looked at that code 🤔

**Comment by IxxyDev** [comment]
Is that free to take?

I'd like to pick this up and work on fixing the remaining cases. 
I'll start with the false positives on optional parameters with non-primitive types and the missing "always falsy" detection, then move on to the other open issues listed in the comments.

**Comment by dyc3** [comment]
sure, @IxxyDev give it a shot

**Comment by IxxyDev** [comment]
@dyc3 @ematipico 

Made a quick research on this, here's the plan and my question

**In scope — bug fixes to the existing rule:**
  - Type-aware detection for optional chaining `(arg?.length with arg: string)` — currently literal-only.
  - Type-aware detection for comparisons, including `x: undefined | string` compared with `null` (per @joprice, confirmed useful by @arendjr).
  - Type-aware detection for logical expressions on member access and call expressions (`arg.test || []` where `arg.test: string[]`).
  - Falsy detection for `!expr` (currently only truthy side is handled for arrays/objects).
  - Fix switch over exhaustive literal unions (see design question below).
  - A related type-inference fix in `biome_js_type_info`: optional params with non-primitive types (`user?: { name: string }, args?: Args`) currently lose their `| undefined` for conditional semantics, causing false positives.

**Out of scope (follow-up issues)**:
  - Array predicate callbacks (`.filter(t => t)`, example #4 from the issue body) — this is a new detection class, not a bug fix, and I'd prefer to scope it separately so the refactor can be reviewed on its own. Happy to open the follow-up as part of this work.
  - `allowConstantLoopConditions` option and `switch (true)` policy — these are configuration decisions that deserve their own discussion.
  - The mutating-variable case from @eliseumds — I suspect this is a flow-analysis limitation in `biome_js_type_info` rather than a rule bug. I'll investigate and file a separate issue if the root cause is in type inference.

Design question on switch: the current implementation routes `JsSwitchStatement` through the generic truthy/falsy check which is why `switch (myValue)` over `'a' | 'b' | 'c'` is flagged — the discriminant has no falsy inhabitants. I don't think the discriminant is the right thing to flag here at all; instead, I'd like to move the check to individual case expressions that are statically unreachable given prior narrowing, leaving the discriminant alone. Does this match your expectations, or do you have a different mental model?

**Comment by dyc3** [comment]
What does the source rule do about it? ideally we stick to what the source rule does.

**Comment by arendjr** [comment]
Thanks for the investigation, @IxxyDev !

It was my impression that detection of optional chaining _should_ already work even for non-literals. But it's possible there could be some bug there of course.

> Design question on switch: the current implementation routes JsSwitchStatement through the generic truthy/falsy check which is why switch (myValue) over 'a' | 'b' | 'c' is flagged — the discriminant has no falsy inhabitants. I don't think the discriminant is the right thing to flag here at all; instead, I'd like to move the check to individual case expressions that are statically unreachable given prior narrowing, leaving the discriminant alone. Does this match your expectations, or do you have a different mental model?

I think your understanding is right. The truthy/falsy check shouldn't be applied to the discriminant only, but for each `<discriminant> == <case>` pair (maybe need to check with the spec whether the semantics should follow the behaviour of the `==` or the `===` operator in `case` expressions... I don't know by head).
